File: bumperbot_utils/bumperbot_utils/safety_stop.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from twist_mux_msgs.action import JoyTurbo
from sensor_msgs.msg import LaserScan
from std_msgs.msg import Bool
from enum import Enum
import math

class State(Enum):
    FREE = 0,
    WARNING = 1, 
    DANGER = 2

class SafetyStop(Node):
    def __init__(self):
        super().__init__("safety_stop")

        self.declare_parameter("danger_distance", 0.2)
        self.declare_parameter("warning_distance", 0.6)
        self.declare_parameter("scan_topic", "scan")
        self.declare_parameter("safety_stop_topic", "safety_stop")
        
        self.danger_distance = self.get_parameter("danger_distance").get_parameter_value().double_value
        self.warning_distance = self.get_parameter("warning_distance").get_parameter_value().double_value
        self.scan_topic = self.get_parameter("scan_topic").get_parameter_value().string_value
        self.safety_stop_topic = self.get_parameter("safety_stop_topic").get_parameter_value().string_value

        self.laser_sub_ = self.create_subscription(LaserScan, 
                                                   self.scan_topic, 
                                                   self.laserCallback,
                                                   10)
        self.safety_stop_pub = self.create_publisher(Bool, 
                                                     self.safety_stop_topic, 
                                                     10)
        
        self.decrease_speed_client = ActionClient(self, 
                                                  JoyTurbo, 
                                                  "joy_turbo_decrease")
        self.increase_speed_client = ActionClient(self, 
                                                  JoyTurbo, 
                                                  "joy_turbo_increase")
        
        while not self.decrease_speed_client.wait_for_server(timeout_sec=1.0) and rclpy.ok():
            self.get_logger().warn("/joy_turbo_decrease is not available waiting...")
        
        while not self.increase_speed_client.wait_for_server(timeout_sec=1.0) and rclpy.ok():
            self.get_logger().warn("/joy_turbo_increase is not available waiting...")
        
        

        
        self.state = State.FREE
        self.prev_state = State.FREE

    def laserCallback(self, msg: LaserScan):
        self.state = State.FREE

        for range_value in msg.ranges:
            if not math.isinf(range_value) and range_value <= self.warning_distance:
                self.state = State.WARNING
                if range_value <= self.warning_distance:
                    self.state = State.DANGER
                    break

        if self.state != self.prev_state:
            is_safety_stop = Bool()
            if self.state == State.WARNING: 
                is_safety_stop.data = False
                self.decrease_speed_client.send_goal_async(JoyTurbo.Goal())
            if self.state == State.FREE:
                is_safety_stop.data = False
                self.increase_speed_client.send_goal_async(JoyTurbo.Goal())
            elif self.state == State.DANGER:
                is_safety_stop.data = True
            self.prev_state = self.state
            self.safety_stop_pub.publish(is_safety_stop)


def main():
    rclpy.init()
    node = SafetyStop()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()
        

File: bumperbot_utils/src/safety_stop.cpp
#include "bumperbot_utils/safety_stop.hpp"
#include <sensor_msgs/msg/laser_scan.hpp>
#include <math.h>


using std::placeholders::_1;

SafetyStop::SafetyStop():
    Node("safety_stop")

{
    declare_parameter<double>("danger_distance", 0.2);
    declare_parameter<std::string>("scan_topic", "scan");
    declare_parameter<std::string>("safety_stop_topic", "safety_stop");

    danger_distance_ = get_parameter("danger_distance").as_double();
    std::string scan_topic = get_parameter("scan_topic").as_string();
    std::string safety_stop_topic = get_parameter("safety_stop_topic").as_string();

    laser_sub_ = create_subscription<sensor_msgs::msg::LaserScan>(
        scan_topic, 
        10, 
        std::bind(&SafetyStop::laserCallback, this, _1)
    );

    

    safety_stop_pub = create_publisher<std_msgs::msg::Bool>(
        safety_stop_topic, 
        10
    );

    state = State::FREE;
}

void SafetyStop::laserCallback( const sensor_msgs::msg::LaserScan & msg)
{
    state = State::FREE;

    for(const auto & range: msg.ranges){
        if(!std::isinf(range) && range <= danger_distance_){
            state = State::DANGER;
            break;
        }
    }

    std_msgs::msg::Bool is_safety_stop;
    if(state == State::DANGER){
        is_safety_stop.data = true;
    }else if (state == State::FREE){
        is_safety_stop.data = false;
    }

    safety_stop_pub->publish(is_safety_stop);

}


int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SafetyStop>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}